lib/asciidoctor/converter.rb

```diff
@@ -1,225 +1,394 @@
-    def composed owner
-    end
-=end
+  # Internal: Mixes the {Config} module into any class that includes the {Converter} module. Additionally, mixes the
+  # {BackendTraits} method into instances of this class.
+  #
+  # into - The Class into which the {Converter} module is being included.
+  #
+  # Returns nothing.
+  private_class_method def self.included into
+    into.include BackendTraits
+    into.extend Config
+  end
 
-    # Public: Converts an {AbstractNode} using the specified transform along
-    # with additional options. If a transform is not specified, implementations
-    # typically derive one from the {AbstractNode#node_name} property.
-    #
-    # Implementations are free to decide how to carry out the conversion. In
-    # the case of the built-in converters, the tranform value is used to
-    # dispatch to a handler method. The {TemplateConverter} uses the value of
-    # the transform to select a template to render.
+  # An abstract base class for defining converters that can be used to convert {AbstractNode} objects in a parsed
+  # AsciiDoc document to a backend format such as HTML or DocBook.
+  class Base
+    include Converter, Logging
+
+    # Public: Converts an {AbstractNode} by delegating to a method that matches the transform value.
     #
-    # node      - The concrete instance of AbstractNode to convert
-    # transform - An optional String transform that hints at which transformation
-    #             should be applied to this node. If a transform is not specified,
-    #             the transform is typically derived from the value of the
-    #             node's node_name property. (optional, default: nil)
-    # opts      - An optional Hash of options that provide additional hints about
-    #             how to convert the node. (optional, default: nil)
+    # This method looks for a method to dispatch to that matches the name of the transform. If the +opts+ argument is
+    # non-nil, this method assumes the dispatch method accepts two arguments, the node and an options Hash. The options
+    # Hash is currently used for the outline transform, but may be used by converters to delegate additional utilities
+    # back to the top-level converter. If the +opts+ argument is nil, this method assumes the dispatch method accepts the
+    # node as its sole argument. Implementations are free to override this method to provide different behavior.
     #
-    # Returns the [String] result
+    # See {Converter#convert} for details about the arguments and return value.
     def convert node, transform = nil, opts = nil
-      raise ::NotImplementedError
+      opts ? (send transform || node.node_name, node, opts) : (send transform || node.node_name, node)
+    rescue
+      if ::NoMethodError === (ex = $!) && ex.receiver == self && ex.name.to_s == (transform || node.node_name)
+        raise ::NotImplementedError, %(#{self.class} (backend: #{@backend}) must implement the ##{ex.name} method)
```

Comment:
@Mogztter Do you think a missing convert delegate method should be an exception or a warning? (essentially the behavior of _skip_with_warning).

Reply:
It depends :wink: 

In my opinion it's easy to miss a warning message and my personal taste is to "fail fast".
But if I don't expect someone to use a given node (for instance a `listing` node because we agree that we won't use them) then I'd probably prefer a warning. In this case we decide that the converter won't implement the `listing` method so the person who use a `listing` block will (hopefully) notice that the block is missing, read the warning message and use another syntax.
In other words, severity can be relative depending on your context. 

A warning message seems more flexible but we "lower" the severity from error (you just can't continue unless you fix it) to warning. But since we can configure `--failure-level=WARN`, I guess it's better if we issue a warning... :thinking:

Reply:
I'm thinking perhaps an error message is most appropriate. The reason I don't think we should fail is that the purpose of this method is to help simplify the development of a converter. But it's not very development friendly if it keeps failing for every unimplemented node. Once a converter is done, this should never happen. So I think that tells us what we should do here.

Reply:
I agree 👍

Reply:
Just pushed that change. I'm going to start with a warning since that's what we were using before in other converters.